### PR TITLE
⚡ perf(cache): BatchGetItem optimization for config resolution (#298)

### LINKED ISSUE #298: ⚡ BatchGetItem optimization for config resolution
## Problem or Use Case

On cache miss, `_resolve_limits()` makes up to 4 sequential GetItem calls to resolve the four-level config hierarchy:
1. Entity limits (specific resource)
2. Entity limits (`_default_`)
3. Resource defaults
4. System defaults

This sequential pattern adds latency (4 round trips) and makes cold-cache performance worse than necessary. DynamoDB's BatchGetItem can fetch all 4 config items in a single request.

Additionally, there's no visibility into cache performance. Without CloudWatch metrics, we can't:
- Monitor cache hit rates in production
- Measure cold-cache latency impact
- Validate optimization effectiveness

### BatchGetItem Optimization

Replace sequential GetItem calls with a single BatchGetItem that fetches all config levels:

```python
# Current: Sequential (4 round trips on cache miss)
entity_limits = await repo.get_limits(entity_id, resource)
entity_default_limits = await repo.get_limits(entity_id, "_default_")
resource_limits = await repo.get_resource_defaults(resource)
system_limits = await repo.get_system_defaults()

# Proposed: Batched (1 round trip)
configs = await repo.batch_get_configs(entity_id, resource)
# Returns all 4 levels in single request
```

Keys to fetch in BatchGetItem:
- `PK=ENTITY#{entity_id}, SK=#CONFIG#{resource}` (entity config)
- `PK=ENTITY#{entity_id}, SK=#CONFIG#_default_` (entity default)
- `PK=RESOURCE#{resource}, SK=#CONFIG` (resource config)
- `PK=SYSTEM#, SK=#CONFIG` (system config)

### CloudWatch Metrics Integration

Publish custom metrics to measure cache effectiveness:

| Metric | Unit | Description |
|--------|------|-------------|
| `ConfigCache/HitRate` | Percent | Hits / (Hits + Misses) over period |
| `ConfigCache/ColdCacheLatency` | Milliseconds | p50/p95/p99 latency on cache miss |
| `ConfigCache/WarmCacheLatency` | Milliseconds | p50/p95/p99 latency on cache hit |

Dimensions: `StackName`, `Resource` (optional)

## Acceptance Criteria

- [ ] `Repository.batch_get_configs(entity_id, resource)` method fetches all 4 config levels in single BatchGetItem call
- [ ] `ConfigCache` falls back to sequential GetItem if BatchGetItem fails (e.g., partial failure)
- [ ] CloudWatch custom metric `ConfigCache/HitRate` published with `StackName` dimension
- [ ] CloudWatch custom metric `ConfigCache/ColdCacheLatency` published with p50/p95/p99 statistics
- [ ] Benchmark tests in `tests/benchmark/` measure cold-cache latency before and after optimization
- [ ] CLAUDE.md updated with BatchGetItem access pattern for config resolution

## Alternatives Considered

1. **Parallel GetItem with asyncio.gather**: Still 4 requests, but concurrent. Lower latency than sequential but higher DynamoDB cost than BatchGetItem.

2. **Eager cache warming**: Pre-fetch all configs on first access. Increases startup latency and may fetch unused configs.

3. **Single composite config item**: Store all config levels in one item. Breaks single-table design and complicates config updates.

### LINKED ISSUE #325: ♻️ Tighten on_unavailable type from str to Literal["allow", "block"] in repository and config_cache layers
## Description

The `on_unavailable` parameter is typed as `str | None` throughout the repository, repository_protocol, and config_cache layers. Meanwhile, the limiter layer already constrains this to the `OnUnavailable` enum (which has exactly two values: `"allow"` and `"block"`). This creates a type gap where the lower layers accept any string, but only two values are semantically valid.

Tighten the type annotation from `str | None` to `Literal["allow", "block"] | None` across these layers to catch invalid values at type-check time rather than runtime.

## Files to Update

**Source files (async):**
- `repository_protocol.py` — `set_system_defaults(on_unavailable: str | None)` parameter and `get_system_defaults() -> tuple[..., str | None]` return type
- `repository.py` — `set_system_defaults(on_unavailable: str | None)` parameter, `get_system_defaults()` return type, and internal deserialization annotations
- `config_cache.py` — All `str | None` annotations for `on_unavailable` in `get_system_defaults()`, `resolve_limits_batched()`, `_process_fetched_items()`, `_evaluate_hierarchy()`, `_resolve_uncached()`, `_evaluate_uncached()`, and related `Callable` type parameters

**Generated files (sync — regenerated via `hatch run generate-sync`):**
- `sync_repository_protocol.py`
- `sync_repository.py`
- `sync_config_cache.py`

**Migration file:**
- `migrations/m_0_8_0_composite_limits.py` — internal `on_unavailable: str | None` annotation

## Acceptance Criteria

- [ ] `repository_protocol.py`: `set_system_defaults` parameter typed as `Literal["allow", "block"] | None`
- [ ] `repository_protocol.py`: `get_system_defaults` return type uses `Literal["allow", "block"] | None`
- [ ] `repository.py`: `set_system_defaults` parameter typed as `Literal["allow", "block"] | None`
- [ ] `repository.py`: `get_system_defaults` return type uses `Literal["allow", "block"] | None`
- [ ] `repository.py`: Internal deserialization annotations use `Literal["allow", "block"] | None`
- [ ] `config_cache.py`: All `on_unavailable` annotations use `Literal["allow", "block"] | None`
- [ ] `config_cache.py`: `Callable` type parameters for batch fetch functions updated
- [ ] `migrations/m_0_8_0_composite_limits.py`: Internal annotation updated
- [ ] Sync files regenerated via `hatch run generate-sync` and up to date
- [ ] `uv run mypy src/zae_limiter` passes with no new errors
- [ ] `uv run pytest tests/unit/` passes with no failures
- [ ] No changes to CLI layer (`cli.py` correctly uses `str` from Click and passes through)
- [ ] No changes to limiter layer (`limiter.py` already uses `OnUnavailable` enum)

## Alternatives Considered

**Use `OnUnavailable` enum in repository layer:** Rejected because the repository layer is intentionally decoupled from the limiter's enum. `Literal["allow", "block"]` provides type safety without introducing a dependency on the limiter module.

**Leave as `str`:** The current state. Works but allows invalid values to pass type checking silently.

### PR DESCRIPTION
**Pull request by sodre**
## Summary
- Add `Repository.batch_get_configs()` method that fetches all 4 config hierarchy levels (entity, entity default, resource, system) in a single `BatchGetItem` call instead of 4 sequential `GetItem` calls
- Integrate batch config resolution into `ConfigCache` with fallback to sequential `GetItem` on partial failure
- Add `ConfigSource` and `OnUnavailableAction` Literal type aliases for type-safe config resolution
- Add benchmark tests to track cold-cache config resolution RCU/latency

## Benchmark Results: BatchGetItem vs Sequential GetItem (#298)

Cold-cache config resolution (moto, no config cache, unique entity IDs per iteration):

| Approach | Mean | Median | OPS | Relative |
|----------|------|--------|-----|----------|
| **Batched** (1 BatchGetItem) | 5.25ms | 5.18ms | 190/s | 1.0x |
| **Sequential** (4 GetItem) | 8.26ms | 8.14ms | 121/s | **1.57x slower** |

### API Call Reduction

| Path | Config Reads | Total BatchGetItem |
|------|-------------|--------------------|
| Before (sequential) | 4 GetItem (entity, entity_default, resource, system) | 1 (buckets) |
| After (batched) | 1 BatchGetItem (4 config keys) | 2 (configs + buckets) |
| Version check | 1 GetItem | unchanged |

**Note:** With moto (in-process mock), the improvement is ~37% because there's no real network latency. On real DynamoDB, eliminating 3 network round trips would yield a significantly larger improvement proportional to network RTT (typically 2-5ms per call).

### Benchmark command

```bash
uv run pytest tests/benchmark/test_operations.py -o "addopts=" -v --benchmark-only -k "config_resolution"
```

## Test plan
- [x] Unit tests for `batch_get_configs()` in `test_repository.py` and `test_sync_repository.py`
- [x] Unit tests for batch config cache integration in `test_config_cache.py` and `test_sync_config_cache.py`
- [x] Edge case coverage: partial results, empty responses, deserialization
- [x] Benchmark test in `test_capacity.py` for cold-cache config resolution
- [x] Verify `uv run pytest tests/unit/ -v` passes (1730 tests)
- [x] Verify `uv run mypy src/zae_limiter` passes

Closes #298
Closes #325

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

**Comment by sodre** [comment]
## Benchmark Results: BatchGetItem vs Sequential GetItem (#298)

Cold-cache config resolution (moto, no config cache, unique entity IDs per iteration):

| Approach | Mean | Median | OPS | Relative |
|----------|------|--------|-----|----------|
| **Batched** (1 BatchGetItem) | 5.25ms | 5.18ms | 190/s | 1.0x |
| **Sequential** (4 GetItem) | 8.26ms | 8.14ms | 121/s | **1.57x slower** |

### API Call Reduction

| Path | Config Reads | Total BatchGetItem |
|------|-------------|--------------------|
| Before (sequential) | 4 GetItem (entity, entity_default, resource, system) | 1 (buckets) |
| After (batched) | 1 BatchGetItem (4 config keys) | 2 (configs + buckets) |
| Version check | 1 GetItem | unchanged |

**Note:** With moto (in-process mock), the improvement is ~37% because there's no real network latency. On real DynamoDB, eliminating 3 network round trips would yield a significantly larger improvement proportional to network RTT (typically 2-5ms per call).

### Benchmark command

```bash
uv run pytest tests/benchmark/test_operations.py -o "addopts=" -v --benchmark-only -k "config_resolution"
```

File: src/zae_limiter/repository_protocol.py [hunk 1]
```diff
@@ -265,6 +265,26 @@ async def get_or_create_bucket(
...
+
+        Args:
+            keys: List of (PK, SK) tuples identifying config items
+
+        Returns:
+            Dict mapping (PK, SK) to (limits, on_unavailable) tuples.
+            on_unavailable is extracted from system config items (None for others).
+            Missing items are not included in the result.
+        """
+        ...
```
[the remark this answers is not in the document]
Comment (sodre) [reply src/zae_limiter/repository_protocol.py]: Dismissing — `...` is the standard Python convention for Protocol method bodies. These are abstract interface declarations, not concrete implementations.

File: tests/unit/test_sync_config_cache.py [hunk 2]
```diff
@@ -399,3 +399,196 @@ def test_no_config_sentinel_is_unique(self) -> None:
...
+            return {(schema.pk_system(), schema.sk_config()): (system_limits, None)}
+
+        limits1, _, source1 = cache.resolve_limits("user-1", "gpt-4", batch_fn)
+        assert limits1 is not None
+        assert source1 == "system"
+        assert call_count == 1
+        limits2, _, source2 = cache.resolve_limits("user-1", "gpt-4", batch_fn)
+        assert limits2 is not None
+        assert source2 == "system"
+        assert call_count == 1
```
[the remark this answers is not in the document]
Comment (sodre) [reply tests/unit/test_sync_config_cache.py]: Dismissing — same false positive as above. This is also a generated sync file (`test_sync_config_cache.py`).
